worlds/stick_ranger/Rules.py

```diff
@@ -0,0 +1,221 @@
+from typing import TYPE_CHECKING, Callable, Dict, List
+
+from BaseClasses import CollectionState, MultiWorld
+from worlds.generic.Rules import set_rule
+
+from .Constants import GOAL_EXIT_LOCATIONS, RANGER_CLASSES
+from .Items import unlocks_by_region
+from .Options import SROptions
+
+if TYPE_CHECKING:
+    from . import StickRanger
+
+
+def class_count(state, player: int) -> int:
+    """Return the number of ranger classes the player has unlocked."""
+    return sum(state.has(f"Unlock {cls} Class", player) for cls in RANGER_CLASSES)
+
+
+def reached_castle(
+    player: int, threshold: int, req_classes: int
+) -> Callable[[CollectionState], bool]:
+    """Return whether or not the player has unlocked access to the Castle stage yet."""
+    return lambda state, _pl=player, _T=threshold, _keys=unlocks_by_region[
+        "Grassland"
+    ]: (
+        state.has("Unlock Castle", _pl, 1)
+        and sum(1 for k in _keys if state.has(k, _pl, 1)) >= _T
+        and class_count(state, _pl) >= req_classes
+    )
```

Comment:
Consider just using `state.can_reach_region` instead of this and the similar helpers below.